File: leetcode/binary_tree/94.binary_tree_inorder_traversal/94.BinaryTreeInorderTraversal_henrytine.py
# Source : https://leetcode.com/problems/binary-tree-inorder-traversal/
# Author : henrytine
# Date   : 2020-08-18

##################################################################################################### 
#
# Given a binary tree, return the inorder traversal of its nodes' values.
# 
# Example:
# 
# Input: [1,null,2,3]
#    1
#     \
#      2
#     /
#    3
# 
# Output: [1,3,2]
# 
# Follow up: Recursive solution is trivial, could you do it iteratively?
#####################################################################################################

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        self.helper(root, res)
        return res

    def helper(self, root, res):
        if root:
            self.helper(root.left, res)
            res.append(root.val)
            self.helper(root.right, res)

# iterator


class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res, stack = [], []

        while stack or root:
            if root:
                stack.append(root)
                root = root.left
            else:
                node = stack.pop()
                res.append(node.val)
                root = node.right
        return res

File: leetcode/binary_tree/94.binary_tree_inorder_traversal/94.BinaryTreeInorderTraversal_henrytine.cpp
// Source : https://leetcode.com/problems/binary-tree-inorder-traversal/
// Author : henrytine
// Date   : 2020-08-29

/***************************************************************************************************** 
 *
 * Given a binary tree, return the inorder traversal of its nodes' values.
 * 
 * Example:
 * 
 * Input: [1,null,2,3]
 *    1
 *     \
 *      2
 *     /
 *    3
 * 
 * Output: [1,3,2]
 * 
 * Follow up: Recursive solution is trivial, could you do it iteratively?
 ******************************************************************************************************/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        helper(root, res);
        return res;
    }
private:
    void helper(TreeNode* node, vector<int> &res) {
         if (!node) return;
         if (node->left) 
             helper(node->left,res);
         res.push_back(node->val);
           
        if (node->right)
            helper(node->right,res);
    }
};


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> res;
        stack<TreeNode*> s;
        
        while (root || !s.empty()) {
            while (root) {
                s.push(root);
                root = root->left; 
            }
            if (!s.empty()) {
                root = s.top();
                s.pop();
                res.push_back(root->val);
                root = root->right;
            }
            
        }
        return res;
    }
};